daemons/attrd/attrd_sync.c

```diff
@@ -0,0 +1,153 @@
+static GHashTable *waitlist = NULL;
+static int waitlist_client = 0;
+
+struct waitlist_node {
+    /* What kind of sync point does this node describe? */
+    enum attrd_sync_point sync_point;
+
+    /* What attribute are we waiting on a sync point for? */
+    char *attr_name;
+
+    /* Information required to construct and send a reply to the client. */
+    char *client_id;
+    uint32_t ipc_id;
+    uint32_t flags;
+};
+
+static void
+next_key(void)
+{
+    do {
+        waitlist_client++;
+        if (waitlist_client < 0) {
+            waitlist_client = 1;
+        }
+    } while (g_hash_table_contains(waitlist, GINT_TO_POINTER(waitlist_client)));
+}
+
+static void
+free_waitlist_node(gpointer data)
+{
+    struct waitlist_node *wl = (struct waitlist_node *) data;
+
+    free(wl->client_id);
+    free(wl->attr_name);
+    free(wl);
+}
+
+static const char *
+sync_point_str(enum attrd_sync_point sync_point)
+{
+    if (sync_point == attrd_sync_point_local) {
+        return "local";
+    } else if  (sync_point == attrd_sync_point_cluster) {
+        return "all nodes";
+    } else {
+        return "unknown";
+    }
+}
+
+void
+attrd_add_client_to_waitlist(pcmk__request_t *request)
+{
+    const char *sync_point = crm_element_value(request->xml, PCMK__XA_ATTR_SYNC_POINT);
+    const char *attr_name = crm_element_value(request->xml, PCMK__XA_ATTR_NAME);
+    struct waitlist_node *wl = NULL;
+
+    if (sync_point == NULL) {
+        return;
+    }
+
+    if (waitlist == NULL) {
+        waitlist = pcmk__intkey_table(free_waitlist_node);
+    }
+
+    wl = calloc(sizeof(struct waitlist_node), 1);
+
+    CRM_ASSERT(wl != NULL);
+
+    wl->client_id = strdup(request->ipc_client->id);
+
+    CRM_ASSERT(wl->client_id);
+
+    if (pcmk__str_eq(sync_point, PCMK__VALUE_LOCAL, pcmk__str_none)) {
+        wl->sync_point = attrd_sync_point_local;
+    } else if (pcmk__str_eq(sync_point, PCMK__VALUE_CLUSTER, pcmk__str_none)) {
+        wl->sync_point = attrd_sync_point_cluster;
+    } else {
+        free_waitlist_node(wl);
+        return;
+    }
+
+    wl->ipc_id = request->ipc_id;
+    wl->flags = request->flags;
+
+    wl->attr_name = strdup(attr_name);
+
+    CRM_ASSERT(wl->attr_name != NULL);
+
+    crm_debug("Added client %s to waitlist for %s sync point on attribute %s",
+              wl->client_id, sync_point_str(wl->sync_point), wl->attr_name);
+
+    next_key();
+    pcmk__intkey_table_insert(waitlist, waitlist_client, wl);
+}
+
+void
+attrd_ack_waitlist_clients(enum attrd_sync_point sync_point, const char *attr_name)
+{
+    GHashTableIter iter;
+    gpointer value;
+
+    if (waitlist == NULL) {
+        return;
+    }
+
+    g_hash_table_iter_init(&iter, waitlist);
+
+    while (g_hash_table_iter_next(&iter, NULL, &value)) {
+        struct waitlist_node *wl = (struct waitlist_node *) value;
+        pcmk__client_t *client = NULL;
+
+        if (wl->sync_point != sync_point || !pcmk__str_eq(wl->attr_name, attr_name, pcmk__str_none)) {
```

Comment:
How do we know this is the particular client's request? Two clients could have requests in flight at the same time for the same attribute with different values. Also, for a multiple-attribute update, we want to ack only after they're all updated.

Reply:
Yeah, I guess the biggest problem I've been having here is figuring out what uniquely identifies a client request.  I could just assign each request an integer when attrd_add_client_to_waitlist is called, but then at ACK time I need some way of figuring that same request ID out so I can find the right item in the waitlist.

I think I was planning on using callid for this purpose at some point, but I am unsure of how these IDs actually get assigned.  And anyway, will I actually know the callid in update_attr_on_host when I go to send the ACK?

For multiple-attribute updates... I don't think there's any way that attrd_updater can do anything with more than one attribute at a time.  The multiple-attribute handling we did for fail-count/last-fail is all hidden in libraries.  I guess we might want to handle sync points in this case for the future, though?  I haven't really thought about how this could be handled.

Reply:
Actually, maybe I'm overthinking this.  See the attached WIP patch.  I've given this cursory testing and it seems to be pretty good.  It still doesn't handle multiple attributes.